### PR TITLE
Trigger

### PR DESCRIPTION
**Pull request by eedm**
Added external triggering capabilities following scheme discussed in issue #7.

### CODE REVIEW

File: instrumental/drivers/cameras/uc480.py [hunk 1]
```diff
@@ -569,6 +573,50 @@ def _color_mode_string(self):
...
+        """
+        delay_us = 0 if delay is None else int(delay.m_as('us'))
+        ret = lib.is_SetTriggerDelay(self._hcam, delay_us)
+        if ret != IS_SUCCESS:
+            print("Error: failed to set trigger delay")
+            
+    def get_trigger_delay(self):
+        """Returns the trigger delay in microseconds
+        """
+        parm = INT()
```
Comment (natezb) [inline instrumental/drivers/cameras/uc480.py]: Typo here

**Comment by natezb** [comment]
Thanks for the contribution.

Have you tested all the code you've added/changed to make sure it works? I spotted one typo in `get_trigger_delay`, which I noted above.

On Monday I'll take a closer look at the code and test out the changes to verify that my camera (that lacks hardware triggering) still works as expected.

**Comment by eedm** [comment]
Thanks. I missed that typo. Fortunately (or unfortunately) it tested out as-is. I also intend to do some more hardware testing with a scope to check that the delay is correct. If you have time, I'd really like your feedback on these changes before I do too much more testing. One problem I already see is that since triggering modes are bit-packed maybe the trigger_mode property should be a function that unpacks the mode into a tuple like (external,continuous,edge).

**Comment by natezb** [comment]
Yeah, I wasn't thinking straight, you can just get rid of that line; it has no effect anyway because you assign to `param` on the next line.

I think you should break out the trigger mode into separate properties for the triggering mode and the edge type. Beyond that, I haven't really fully decided how we should implement camera settings. For instance, using read/write properties instead of `set_x()` and `get_x()` methods is certainly more pythonic, but in some cases obscures the fact that you may be communicating with hardware. One thing I've been doing (check out `pco.py` and `pixelfly.py`, as well as `__init__.py` in the `cameras` directory) is using `DEFAULT_KWDS` to handle these sorts of things. The idea is that, when calling `start_live_video()`, `grab_image()`, and others, the user may pass in keyword arguments that apply only to that capture. If a keyword is not provided, the camera uses a default setting, which can be configured via the `set_defaults()` method available to any camera object.

At this point I'm not convinced of the best course, so I'd say to do whatever you're comfortable with. But yeah, definitely split off/software/hardware and rising/falling into separate properties.

And just a few comments on your code:
- If you encounter an error, you should raise an exception instead of printing it. I think my older version of this module had a bad habit of printing errors, but I've since cleared that up
- Thanks for using docstrings in your methods. Make sure that you're using [numpydoc format](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) so that sphinx can parse and render it correctly (e.g. the parameters list in the docstring for `set_trigger()`.

**Comment by eedm** [comment]
I'm closing this pull request to work on the code comments above: exceptions, rather than prints, and using numpydoc.

**Pull request closed without merging**